pkg/virtctl/vmexport/vmexport.go

```diff
@@ -0,0 +1,623 @@
+	if err != nil {
+		return err
+	}
+	if vmexport == nil {
+		return fmt.Errorf("Unable to get '%s/%s' VirtualMachineExport", namespace, vmexportName)
+	}
+
+	// Download the exported volume
+	if err := downloadVolume(client, vmexport); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// downloadVolume handles the process of downloading the requested volume from a VirtualMachineExport
+func downloadVolume(client kubecli.KubevirtClient, vmexport *exportv1.VirtualMachineExport) error {
+	output, err := os.Create(outputFile)
+	if err != nil {
+		return err
+	}
+	defer util.CloseIOAndCheckErr(output, nil)
+
+	// Extract the URL from the vmexport
+	downloadUrl, err := getUrlFromVirtualMachineExport(client, vmexport)
+	if err != nil {
+		return err
+	}
+
+	resp, err := handleHTTPRequest(client, vmexport, downloadUrl)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Check server response
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("Bad status: %s", resp.Status)
+	}
+
+	// Lastly, copy the file to the expected output
+	if err := copyFileWithProgressBar(output, resp); err != nil {
+		return err
+	}
+
+	fmt.Println("Download finished succesfully")
+	return nil
+}
+
+// getUrlFromVirtualMachineExport inspects the VirtualMachineExport status to fetch the extected URL
+func getUrlFromVirtualMachineExport(client kubecli.KubevirtClient, vmexport *exportv1.VirtualMachineExport) (string, error) {
+	var downloadUrl string
+
+	if vmexport.Status.Links.External.Volumes == nil {
+		return "", fmt.Errorf("Unable to access the volume info from '%s/%s' VirtualMachineExport", vmexport.Namespace, vmexport.Name)
+	}
+
+	for _, exportVolume := range vmexport.Status.Links.External.Volumes {
+		// Access the requested volume
+		if exportVolume.Name == volumeName {
+			for _, format := range exportVolume.Formats {
+				// We always attempt to find and get the compressed file URL
+				if format.Format == exportv1.KubeVirtGz || format.Format == exportv1.ArchiveGz {
+					downloadUrl = format.Url
+					break
+				} else if format.Format == exportv1.KubeVirtRaw {
+					downloadUrl = format.Url
+				}
+			}
+		}
+	}
+
+	if downloadUrl == "" {
+		return "", fmt.Errorf("Unable to get a valid URL from '%s/%s' VirtualMachineExport", vmexport.Namespace, vmexport.Name)
+	}
+
+	return downloadUrl, nil
+}
+
+// waitForVirtualMachineExport waits for the VirtualMachineExport status to be ready
+func waitForVirtualMachineExport(client kubecli.KubevirtClient, namespace string, interval, timeout time.Duration) error {
+	err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		vmexport, err := getVirtualMachineExport(client, namespace)
+		if err != nil {
+			return true, err
+		}
+
+		if vmexport == nil {
+			return true, err
+		}
+
+		if vmexport.Status == nil {
+			return false, nil
+		}
+
+		done := vmexport.Status.Phase == "Ready"
+		if !done {
+			fmt.Printf("Waiting for VM Export %s status to be ready...\n", vmexportName)
```

Comment:
do we want this message to be printed every 2 seconds until the export is ready?

Reply:
The external links shouldn't take too long to be ready, so I think it's an appropriate interval. Also, I used the imageupload command as a reference for this section, so it has the same intervals.

Reply:
if in total it shouldn't take too long then why the timeout is 24hours? it seems a lot of print every 2 seconds for 24hours :)

Reply:
I am going to agree that waiting that long is excessive, I would say a minute or two is more than enough. If it is not ready by then, something is going on.